[Develop] Support for RHEL9 and Rocky9

hanwen-pcluste opened the pull request:
### Description of changes
* See commits descriptions for details
* Please ignore the last two [temporary] commits

### Tests
* Kitchen tests on c5n.xlarge,g4dn.2xlarge,m6g.xlarge,c6g.xlarge on all OSes are passed
* Integration tests in https://github.com/aws/aws-parallelcluster/blob/develop/tests/integration-tests/configs/new_os.yaml have been run. No failures related to this PR.
### References
* CLI change: https://github.com/aws/aws-parallelcluster/pull/6018

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hanwen-cluster committed:
Initial code change to add RHEL9 and Rock9

The initial tedious changes to add RHEL9 and Rock9 to Chef resources and kitchen tests

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Packages installation changes for RHEL8, RHEL9, Rocky8, Rocky9

Remove `pyparted`, `redhat-lsb`, `python2`, `python2-pip` installations because they are not available in RHEL9 and Rocky9.

Add perl, dbus-devel installation because they were part of `redhat-lsb` and are dependencies of Slurm

We are doing the same on RHEL8, RHEL9, Rocky8, Rocky9 to keep consistencies across OSes.

Reference:
pyparted: https://github.com/dcantrell/pyparted
redhat-lsb: https://access.redhat.com/solutions/6973382
perl: https://www.perl.org
dbus-devel: https://packages.fedoraproject.org/pkgs/dbus/dbus-devel/

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Simplify logic around udev rules

ParallelCluster uses udev to trigger a Python script upon EBS attachment. Specifically, when it detects an attach from a device (e.g. /dev/xvdb), ParallelCluster udev rule creates a symbolic link from `/dev/disk/by-ebs-volumeid` (e.g. `/dev/disk/by-ebs-volumeid/vol-123456`) to the device name (/dev/xvdb). Then, our cookbook checks the device under `/dev/disk/by-ebs-volumeid` is ready.

Prior to this commit, ParallelCluster udev script used boto3 calls to retrieve volume id from device name.

Seems starting from RHEL 9, the scripts triggered by udev no long have network access for security reasons. Therefore, this commit removes the boto3 calls and get the volume id from `/dev/disk/by-ebs-volumeid/parallelcluster_dev_id_mapping`, which is a file ParallelCluster cookbook writes to. Although the logic could be further simplified by not using the udev rule at all, this commit takes the first step of improvement without spending too much time to reach the absolute simplicity.

Moreover, prior to this commit, a service was created to trigger `udevadm trigger`. `udevadm trigger` is used to refresh udev system after udev rules changes. It is unnecessary to have a service trigger the command. Therefore, this commit deletes the service and triggers the command once after the rules changes.

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Remove unused Munge SysV init logic

The Munge installation is already based on systemd. The SysV Init logic is no longer used. We forgot to remove this logic when we dropped support for Amazon Linux 1

When running kitchen tests on Docker, munge on some OSes still uses SysV. Therefore, we skip some `service` resource execution. We take the kitchen tests on EC2 instance as the golden standard. This is consistent with what we've been doing (i.e. We have been skip some `service` resource for other packages)

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Convert IMDS lockdown from SysV init to systemd

`test_create_imds_secured` integration test has been passed

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Upgrade ARM PL to 23.10

Different GCC versions are used for different operating systems. For existing operating systems, we keep the version unchanged (9.3 for alinux2, centos7, ubuntu20, rhel8, rocky8; 11.3 for ubuntu22). For newly added OSes (RHEL9 and Rocky9), we use 11.3 because 9.3 is not available.

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-cluster committed:
Configure c states using `grubby` command

For other OSes, this commit keeps the old approach of changing grub files. This is done because the old approach does not take effect on RHEL9 and Rocky9. The `grubby` command does not take effect on some of the old OSes.

Signed-off-by: Hanwen <hanwenli@amazon.com>

gmarciani commented on cookbooks/aws-parallelcluster-environment/recipes/config/imds.rb:
```diff
@@ -60,8 +60,8 @@
 
-  template '/etc/init.d/parallelcluster-iptables' do
-    source 'imds/parallelcluster-iptables.erb'
+  template '/usr/local/sbin/restore_tables.sh' do
```
Why moving away from init.d?

hanwen-pcluste replied:
RHEL is moving away from init.d: https://access.redhat.com/solutions/7030065
The old code does not work on RHEL 9 AMIs

gmarciani commented on cookbooks/aws-parallelcluster-environment/resources/ec2_udev_rules/partial/_common_udev_configuration.rb:
```diff
@@ -67,8 +58,7 @@
-    supports restart: true
-    action %i(enable start)
+  execute "Refresh UdevAdmin" do
+    command "udevadm trigger --action=change --subsystem-match=block"
```
May you please add a brief comment about why this is required?

hanwen-pcluste replied:
I am not 100% sure that this is required. This was done in the old code. Therefore I kept here to stay on the safe side

gmarciani replied:
Let's validate in a follow up activity if it is required or not. Better to not have not required steps

gmarciani commented on cookbooks/aws-parallelcluster-environment/spec/unit/recipes/imds_spec.rb:
```diff
@@ -56,11 +56,6 @@
           is_expected.to run_execute("Save ip6tables rules").with(command: /ip6tables-save/)
         end
 
-        it 'creates iptables init.d file' do
```
Different templates are created as part of this PR (eg: parallelcluster-iptables.service), so it should be covered in this test.

hanwen-pcluste replied:
Done

gmarciani commented on cookbooks/aws-parallelcluster-environment/templates/imds/parallelcluster-iptables.service.erb:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Setup ephemeral drives service
```
This service is not about setting up ephemeral drivers, but about configuring iptables.

hanwen-pcluste replied:
Done

gmarciani commented on cookbooks/aws-parallelcluster-platform/resources/arm_pl/arm_pl_ubuntu22.rb:
```diff
@@ -31,4 +26,8 @@ def armpl_platform
   end
+
+  def gcc_major_minor_version
+    '11.3'
```
Why are we upgrading gcc as part of adding the support for Rocky/RHEL?

hanwen-pcluste replied:
According to the commit description:
> Different GCC versions are used for different operating systems. For existing operating systems, we keep the version unchanged (9.3 for alinux2, centos7, ubuntu20, rhel8, rocky8; 11.3 for ubuntu22). For newly added OSes (RHEL9 and Rocky9), we use 11.3 because 9.3 is not available.

gmarciani commented on cookbooks/aws-parallelcluster-platform/resources/c_states/c_states_redhat9.rb:
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+# Copyright:: 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
HEre and elsewhere: on new files the cipyright year should be 2024.

hanwen-pcluste replied:
Done

gmarciani commented on cookbooks/aws-parallelcluster-platform/resources/install_packages/install_packages_redhat8.rb:
```diff
@@ -23,12 +23,13 @@ def default_packages
   # environment-modules required by EFA, Intel MPI and ARM PL
   # Removed libssh2-devel from base_packages since is not shipped by RedHat 8 and in conflict with package libssh-0.9.6-3.el8.x86_64
   # iptables needed for IMDS setup
+  # ToDo send a message in pcluster-interest to see if the package removal/addition is ok
```
Let's remove this comment about pcluster interest

hanwen-pcluste replied:
Done

gmarciani commented on cookbooks/aws-parallelcluster-platform/spec/unit/resources/nvidia_repo_spec.rb:
```diff
@@ -127,7 +127,7 @@ def self.remove(chef_run)
           'rhel7'
         when 'redhat'
-          'rhel8'
+          "rhel#{version}"
```
Why in thre change above it is required a `.to_i` and here it is now?

hanwen-pcluste replied:
For Ubuntu `.to_i` is required because we want to convert `22.04` to `22`. For RHEL `.to_i` is optional

hgreebe committed:
Update mySQL version

hanwen-cluster committed:
[kitchen tests] Improve Docker mocker

1. /etc/init.d is no long used. Therefore, the mocks are removed
2. Install Python. RHEL9 docker does not come with Python. We ere just lucky that previous docker had Python installed

Signed-off-by: Hanwen <hanwenli@amazon.com>

hanwen-pcluste committed:
Merge branch 'develop' into developforpr
